zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/request/AuthorizationRequest.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    return AuthorizationCheckBehavior.FORBIDDEN_ERROR_MESSAGE_WITH_RESOURCE.formatted(
+        permissionType,
+        resourceType,
+        resourceIds.stream()
+            .filter(resourceId -> resourceId != null && !resourceId.isEmpty())
+            .sorted()
+            .collect(Collectors.joining(", ")));
+  }
+
+  public String getTenantErrorMessage() {
+    final var errorMsg =
+        isNewResource
+            ? AuthorizationCheckBehavior.FORBIDDEN_FOR_TENANT_ERROR_MESSAGE
+            : AuthorizationCheckBehavior.NOT_FOUND_FOR_TENANT_ERROR_MESSAGE;
+    return errorMsg.formatted(permissionType, resourceType, tenantId);
+  }
+
+  public static AuthorizationRequest.Builder builder() {
+    return new AuthorizationRequest.Builder();
+  }
+
+  public static AuthorizationRequest of(final Consumer<Builder> consumer) {
+    final var builder = new AuthorizationRequest.Builder();
+    consumer.accept(builder);
+    return builder.build();
+  }
+
+  public static final class Builder {
+
+    private TypedRecord<?> command;
+    private Map<String, Object> authorizationClaims;
+    private AuthorizationResourceType resourceType;
+    private PermissionType permissionType;
+    private Set<String> resourceIds;
+    private String tenantId;
+    private boolean isNewResource;
+
+    public Builder() {
+      authorizationClaims = new HashMap<>();
+      resourceIds = new HashSet<>();
+    }
+
+    public Builder command(final TypedRecord<?> command) {
+      this.command = command;
+      return this;
+    }
+
+    public Builder authorizationClaims(final Map<String, Object> authorizationClaims) {
+      this.authorizationClaims = authorizationClaims;
+      return this;
+    }
+
+    public Builder resourceType(final AuthorizationResourceType resourceType) {
+      this.resourceType = resourceType;
+      return this;
+    }
+
+    public Builder permissionType(final PermissionType permissionType) {
+      this.permissionType = permissionType;
+      return this;
+    }
+
+    public Builder tenantId(final String tenantId) {
+      this.tenantId = tenantId;
+      return this;
+    }
+
+    public Builder isNewResource(final boolean isNewResource) {
+      this.isNewResource = isNewResource;
+      return this;
+    }
+
+    public Builder newResource() {
+      return isNewResource(true);
+    }
+
+    public Builder addResourceId(final String resourceId) {
+      resourceIds.add(resourceId);
+      return this;
+    }
+
+    public Builder addAllResourceIds(final Collection<String> resourceIds) {
+      resourceIds.forEach(this::addResourceId);
+      return this;
+    }
+
+    public AuthorizationRequest build() {
+      final Map<String, Object> claims =
+          command != null ? command.getAuthorizations() : authorizationClaims;
+      final boolean isTenantOwnedResource = tenantId != null && !tenantId.isEmpty();
+      final boolean triggeredByInternalCommand = command != null && command.isInternalCommand();
```

Comment:
❓ Does this mean if someone does not use the builder, the values may be incorrect? What would happen in this case? Should we be more defensive and either:

1. Not use a builder but do everything in the constructor, or
2. Assert the values in the constructor, or
3. Enforce usage of the builder by making the constructor private.